lib/socket/raw.lua

```diff
@@ -0,0 +1,79 @@
+--
+-- SPDX-FileCopyrightText: (c) 2025 Ashwani Kumar Kamal <ashwanikamal.im421@gmail.com>
+-- SPDX-License-Identifier: MIT OR GPL-2.0-only
+--
+
+---
+-- RAW AF_PACKET socket operations.
+-- This module provides a higher-level abstraction over the `socket` module.
+--
+-- @module socket.raw
+-- @see socket
+--
+local socket = require("socket")
+
+local af   = socket.af
+local sock = socket.sock
+
+local raw = {}
+
+---
+-- Creates and binds a raw packet socket for receiving frames.
+-- @param ifindex (number) Interface index.
+-- @param proto (number) [optional] EtherType (defaults to ETH_P_ALL).
+-- @return A new raw packet socket bound for RX.
+-- @raise Error if socket.new() or socket.bind() fail.
+-- @see socket.new
+-- @see socket.bind
+function raw.bind(ifindex, proto)
+	proto = proto or socket.ETH_P_ALL
+
+	local s = socket.new(af.PACKET, sock.RAW, proto)
+	s:bind(string.pack(">H", proto), ifindex)
+	return s
+end
+
+---
+-- Creates and binds a raw packet socket for transmitting frames.
+-- @param ifindex (number) Interface index.
+-- @param proto (number) EtherType.
+-- @return A new raw packet socket bound for TX.
+-- @raise Error if socket.new() or socket.bind() fail.
+-- @see socket.new
+-- @see socket.bind
+function raw.connect(ifindex, proto)
+	assert(proto, "raw.connect: protocol required")
+
+	local s = socket.new(af.PACKET, sock.RAW, proto)
+	s:bind(ifindex)
+	return s
+end
+
+---
+-- Creates a scoped receiving raw socket
+-- @param ifindex (number) Interface index.
+-- @param proto (number) EtherType.
+-- @param fn (function) procedure to execute when receiving frames
+-- @return Outout of raw packet socket bound for RX.
+-- @raise Error if socket.raw.bind() fail.
+-- @see socket.raw.bind
+function raw.with_rx(ifindex, proto, fn)
+    local s <close> = raw.bind(ifindex, proto)
+    return fn(s)
+end
+
+---
+-- Creates a scoped transmitting raw socket
+-- @param ifindex (number) Interface index.
+-- @param proto (number) EtherType.
+-- @param fn (function) procedure to execute when transmitting frames
+-- @return Outout of raw packet socket bound for TX.
+-- @raise Error if socket.raw.connect() fail.
+-- @see socket.raw.connect
+function raw.with_tx(ifindex, proto, fn)
+    local s <close> = raw.connect(ifindex, proto)
+    return fn(s)
+end
```

Comment:
I don't think we need this.. I'd rather keep it simple.. having only bind and connect